File: v4/rust/iprange-livedb/src/recovery/indirect_build.rs
//! Shared immutable construction for membership-backed recovery modes.

use crate::cancellation::CancellationToken;
use crate::contract::{AddressFamily, MetaV4, ValueKind};
use crate::error::{Error, Result};
use crate::immutable_output::Builder;
use crate::key::{Ipv4Key, Ipv6Key};
use crate::mapping::Mapping;

use super::construction::{self, Construction, Failure};
use super::membership::{analyze, IndirectAnalysis};
use super::membership_index::MembershipIndex;
use super::membership_output::MembershipKey;
use super::page_set::PageSet;
use super::range_build::{
    failed_pages, retained_metadata_bytes, BuildResult, RangeBuild, SortReuse,
};
use super::report::{RecoveryReport, RecoverySink, Reporter};
use super::structure_table::StructureIndex;
use super::structured_output::StructuredKey;
use super::tables::Tables;
use super::RecoveryBudget;

pub(super) struct OutputContext<'a, 'b, 'c, S> {
    pub(super) request: RangeBuild<'a>,
    pub(super) pages: PageSet,
    pub(super) memberships: &'a MembershipIndex,
    pub(super) structures: Option<&'a StructureIndex>,
    pub(super) tables: &'a Tables,
    pub(super) builder: &'b mut Builder,
    pub(super) reporter: &'b mut Reporter<'c, S>,
}

pub(super) trait Mode {
    const VALUE_KIND: ValueKind;

    fn check_structures(structures: Option<&StructureIndex>) -> Result<()>;

    #[allow(clippy::result_large_err)]
    fn output<K, S>(context: OutputContext<'_, '_, '_, S>) -> BuildResult
    where
        K: MembershipKey + StructuredKey,
        S: RecoverySink;
}

#[allow(clippy::result_large_err)]
pub(super) fn construct<M: Mode, S: RecoverySink>(
    mapping: &Mapping,
    source_meta: MetaV4,
    builder: Builder,
    budget: &RecoveryBudget,
    cancellation: &CancellationToken,
    sink: &mut S,
) -> std::result::Result<Construction, Failure> {
    let (builder, analysis) = construction::prepare(builder, source_meta, M::VALUE_KIND, || {
        analyze(
            mapping,
            source_meta,
            budget,
            cancellation,
            sink,
            M::VALUE_KIND,
        )
    })?;
    match source_meta.address_family {
        AddressFamily::Ipv4 => build::<M, Ipv4Key, S>(
            mapping,
            source_meta,
            builder,
            budget,
            cancellation,
            sink,
            analysis,
        ),
        AddressFamily::Ipv6 => build::<M, Ipv6Key, S>(
            mapping,
            source_meta,
            builder,
            budget,
            cancellation,
            sink,
            analysis,
        ),
    }
}

#[allow(clippy::too_many_arguments, clippy::result_large_err)]
fn build<M, K, S>(
    mapping: &Mapping,
    source_meta: MetaV4,
    mut builder: Builder,
    budget: &RecoveryBudget,
    cancellation: &CancellationToken,
    sink: &mut S,
    analysis: IndirectAnalysis,
) -> std::result::Result<Construction, Failure>
where
    M: Mode,
    K: MembershipKey + StructuredKey,
    S: RecoverySink,
{
    let IndirectAnalysis {
        report,
        readable_records,
        ordered,
        catalog,
        memberships,
        structures,
        tables,
        metadata,
        pages,
    } = analysis;
    if let Err(cause) = M::check_structures(structures.as_ref()) {
        return Err(failure(builder, pages, cause, report));
    }
    if let Err(cause) =
        catalog.for_each(&tables, |entry| builder.push_feed(entry.name, entry.index))
    {
        return Err(failure(builder, pages, cause, report));
    }
    let retained = match retained_bytes(&tables, &metadata) {
        Ok(retained) => retained,
        Err(cause) => return Err(failure(builder, pages, cause, report)),
    };
    #[cfg(any(unix, windows))]
    let sort_reuse = SortReuse::area(tables.scratch_region());
    #[cfg(not(any(unix, windows)))]
    let sort_reuse = SortReuse::none();
    construction::complete_ranges(
        builder,
        metadata.as_deref(),
        budget.max_heap_bytes,
        retained_metadata_bytes(&metadata),
        report,
        sink,
        move |builder, reporter| {
            let result = M::output::<K, S>(OutputContext {
                request: RangeBuild {
                    mapping,
                    meta: source_meta,
                    budget,
                    cancellation,
                    readable_records,
                    ordered,
                    retained_heap_bytes: retained,
                    sort_reuse,
                },
                pages,
                memberships: &memberships,
                structures: structures.as_ref(),
                tables: &tables,
                builder,
                reporter,
            });
            drop(tables);
            result
        },
    )
}

fn retained_bytes(tables: &Tables, metadata: &Option<Vec<u8>>) -> Result<u64> {
    tables
        .retained_bytes()
        .checked_add(retained_metadata_bytes(metadata))
        .ok_or(Error::ArithmeticOverflow("recovery retained heap"))
}

fn failure(builder: Builder, pages: PageSet, cause: Error, report: RecoveryReport) -> Failure {
    let failed = failed_pages(pages, cause);
    construction::failure(builder, failed.cause, report, failed.scratch)
}

File: v4/go/internal/recovery/indirect_build.go
package recovery

// Shared immutable construction for membership-backed recovery modes
// (Rust recovery/indirect_build.rs): one analysis is prepared against
// the destination builder, the feeds are re-pushed from the recovered
// catalog, and the family build folds the mode output through the
// shared range build and the complete-ranges finish.

import (
	"github.com/firehol/iprange/v4/go/internal/format"
	"github.com/firehol/iprange/v4/go/internal/mapping"
	"github.com/firehol/iprange/v4/go/internal/writer"
)

// indirectMode is one membership-backed recovery output mode (Rust
// Mode): the value-kind guard, the structure proof, and the range
// output policy over the recovery tables.
type indirectMode struct {
	kind            uint8
	checkStructures func(structures *structureIndex) error
	output          func(context indirectOutputContext) (any, *rangeBuildFailure)
}

// indirectOutputContext carries one mode output request (Rust
// OutputContext).
type indirectOutputContext struct {
	request     rangeBuild
	pages       *pageSet
	memberships *membershipIndex
	structures  *structureIndex
	tables      *tableStore
	builder     *writer.OutputBuilder
	rep         *reporter
}

// indirectConstruct runs one membership-backed recovery construction
// (Rust indirect_build::construct): the preparation proves the
// destination against the source generation and runs the analysis,
// then the family build streams the mode output.
func indirectConstruct(mode indirectMode, m *mapping.Mapping, sourceMeta format.Meta, builder *writer.OutputBuilder, budget *RecoveryBudget, check func() error, sink RecoverySink) (*Construction, *constructionFailure) {
	analysis, failure := prepareConstruction(builder, sourceMeta, mode.kind, func() (any, *analysisFailure) {
		result, failed := indirectAnalyze(m, sourceMeta, budget, check, sink, mode.kind)
		if failed != nil {
			return nil, failed
		}
		return result, nil
	})
	if failure != nil {
		return nil, failure
	}
	indirect := analysis.(*indirectAnalysis)
	if err := mode.checkStructures(indirect.structures); err != nil {
		return nil, constructionFailureOf(builder, err, indirect.report, nil)
	}
	codec, ok := indirectCodec(sourceMeta.AddressFamily)
	if !ok {
		return nil, constructionFailureOf(builder, &format.Error{Code: format.CodeFormatInvalid, Detail: "recovery indirect family is invalid"}, indirect.report, nil)
	}
	return indirectBuild(mode, codec, m, sourceMeta, builder, budget, check, sink, indirect)
}

// indirectBuild runs the family build over one completed analysis
// (Rust indirect_build::build: the structure proof, the recovered
// catalog feeds, the retained tables heap, and the complete-ranges
// finish over the mode output).
func indirectBuild(mode indirectMode, codec rangeCodec, m *mapping.Mapping, sourceMeta format.Meta, builder *writer.OutputBuilder, budget *RecoveryBudget, check func() error, sink RecoverySink, analysis *indirectAnalysis) (*Construction, *constructionFailure) {
	if err := analysis.catalog.forEach(analysis.tables, func(entry catalogFeed) error {
		return builder.PushFeed(string(entry.name), entry.index)
	}); err != nil {
		return nil, constructionFailureOf(builder, err, analysis.report, nil)
	}
	retained, err := retainedIndirectBytes(analysis.tables, analysis.metadata)
	if err != nil {
		return nil, constructionFailureOf(builder, err, analysis.report, nil)
	}
	// The metadata compression window charges only the retained
	// metadata bytes (Rust indirect_build::build: complete_ranges over
	// retained_metadata_bytes); the full retained tables heap travels
	// only in the range-build context below.
	return completeRanges(builder, analysis.metadata, budget.MaxHeapBytes, retainedMetadataBytes(analysis.metadata), analysis.report, sink, func(builder *writer.OutputBuilder, rep *reporter) (any, *rangeBuildFailure) {
		context := indirectOutputContext{
			request: rangeBuild{
				mapping:           m,
				meta:              sourceMeta,
				budget:            budget,
				check:             check,
				readableRecords:   analysis.readableRecords,
				ordered:           analysis.ordered,
				retainedHeapBytes: retained,
			},
			pages:       analysis.pages,
			memberships: analysis.memberships,
			structures:  analysis.structures,
			tables:      analysis.tables,
			builder:     builder,
			rep:         rep,
		}
		return mode.output(context)
	})
}

// retainedIndirectBytes is the heap retained by the tables and the
// analyzed metadata (Rust indirect_build::retained_bytes).
func retainedIndirectBytes(tables *tableStore, metadata []byte) (uint64, error) {
	retained, ok := checkedAdd(tables.retainedBytes(), retainedMetadataBytes(metadata))
	if !ok {
		return 0, overflowError("recovery retained heap")
	}
	return retained, nil
}

// indirectCodec returns the family codec of one indirect source.
func indirectCodec(family uint8) (rangeCodec, bool) {
	switch family {
	case format.AddressFamilyIPv4:
		return rangeV4Codec{}, true
	case format.AddressFamilyIPv6:
		return rangeV6Codec{}, true
	default:
		return nil, false
	}
}
